Syntax guide - update re multiple line syntax blocks

hamishwillee opened the pull request:
This attempts to reflect discussion in https://github.com/mdn/content/issues/2202

Specifically syntax sections are now split out multi-line rather than using formal BNF-style notation. I've also made this a little more proscriptive - i.e. "do it this way by preference" rather than "this is a way you can do it"

Note, I have no idea how you're supposed to find the doc https://developer.mozilla.org/en-US/docs/MDN/Structures/Syntax_sections

hamishwillee committed:
Update Syntax sections in writing guide.

Co-authored-by: Florian Scholz <fs@florianscholz.com>

hamishwillee commented on files/en-us/mdn/structures/syntax_sections/index.html:
```diff
@@ -16,11 +16,20 @@
+  &lt;h2 id="Syntax"&gt;Syntax&lt;/h2&gt;
+
+  &lt;pre class="brush: js"&gt;slice()
+  slice(start)
```
Should we return Array ?

Elchi3 replied:
Not for the moment

hamishwillee commented on files/en-us/mdn/structures/syntax_sections/index.html:
```diff
@@ -31,7 +40,7 @@ <h3 id="General_style_rules">General style rules</h3>
   <p>All programmer-specified content in a syntax block should be italicized using the {{HTMLElement("var")}} element:</p>
 
-  <pre class="brush: js notranslate"><var>responseStr</var> = <var>element</var>.querySelector(<var>selector</var>);</pre>
+  <pre class="brush: js"><var>responseStr</var> = <var>element</var>.querySelector(<var>selector</var>);</pre> 
```
See previous line - do we really mark everything up with `var`?

Elchi3 replied:
This example looks quite confusing to me. I'm not sure what to make of it.
Wouldn't this be the correct syntax: `queryselector(selector)` ?

hamishwillee replied:
@Elchi3 I didn't even consider we should roll this out to the single line case. Makes a lot of sense, but just to make sure everyone is looped in I have added https://github.com/mdn/content/issues/2202#issuecomment-781725210 - will come back to this once people have had a chance to review.

Elchi3 commented on files/en-us/mdn/structures/syntax_sections/index.html:
```diff
@@ -16,11 +16,20 @@
+
+  &lt;pre class="brush: js"&gt;slice()
+  slice(start)
+  slice(start,end)
```
```suggestion
  slice(start, end)
```

Elchi3 commented on files/en-us/mdn/structures/syntax_sections/index.html:
```diff
@@ -41,62 +50,73 @@ <h3 id="Constructors_and_methods">Constructors and methods</h3>
+<p>Start with a syntax block, like this (from the {{DOMxRef("IntersectionObserver.IntersectionObserver", "IntersectionObserver constructor")}} page):</p>
 
-<p>or this (see {{DOMxRef("WindowOrWorkerGlobalScope.fetch")}}):</p>
+<pre class="brush: js">var <var>observer</var> = new IntersectionObserver(<var>callback</var>, <var>options</var>);</pre>
```
```suggestion
<pre class="brush: js">new IntersectionObserver(<var>callback</var>, <var>options</var>);</pre>
```

Elchi3 commented on files/en-us/mdn/structures/syntax_sections/index.html:
```diff
@@ -41,62 +50,73 @@ <h3 id="Constructors_and_methods">Constructors and methods</h3>
+<pre class="brush: js">var <var>observer</var> = new IntersectionObserver(<var>callback</var>, <var>options</var>);</pre>
 
-<pre class="brush: js notranslate"><var>promiseResponse</var> = fetch(<var>input</var>, <var>init</var>);</pre>
+<p>or this (from {{DOMxRef("WindowOrWorkerGlobalScope.fetch")}}):</p>
```
Can we avoid using a mixin example here? That just adds complexity that isn't needed to explain this.

Elchi3 commented on files/en-us/mdn/structures/syntax_sections/index.html:
```diff
@@ -41,62 +50,73 @@ <h3 id="Constructors_and_methods">Constructors and methods</h3>
+<p>Similarly, for {{DOMxRef("CanvasRenderingContext2D.drawImage")}}:</p>
 
-<pre class="brush: js notranslate">void <var>ctx</var>.drawImage(<var>image</var>, <var>dx</var>, <var>dy</var>);
+<pre class="brush: js">void <var>ctx</var>.drawImage(<var>image</var>, <var>dx</var>, <var>dy</var>);
 void <var>ctx</var>.drawImage(<var>image</var>, <var>dx</var>, <var>dy</var>, <var>dWidth</var>, <var>dHeight</var>);
 void <var>ctx</var>.drawImage(<var>image</var>, <var>sx</var>, <var>sy</var>, <var>sWidth</var>, <var>sHeight</var>, <var>dx</var>, <var>dy</var>, <var>dWidth</var>, <var>dHeight</var>);</pre>
```
```suggestion
<pre class="brush: js">drawImage(<var>image</var>, <var>dx</var>, <var>dy</var>);
drawImage(<var>image</var>, <var>dx</var>, <var>dy</var>, <var>dWidth</var>, <var>dHeight</var>);
drawImage(<var>image</var>, <var>sx</var>, <var>sy</var>, <var>sWidth</var>, <var>sHeight</var>, <var>dx</var>, <var>dy</var>, <var>dWidth</var>, <var>dHeight</var>);</pre>
```

Elchi3 commented on files/en-us/mdn/structures/syntax_sections/index.html:
```diff
@@ -41,62 +50,73 @@ <h3 id="Constructors_and_methods">Constructors and methods</h3>
+
+<pre class="brush: js">arr.slice([begin[, end]])</pre>
+
+<p>For specific cases where it is seen as beneficial, a separate <strong>Formal syntax</strong> section may be declared using ththe formal notification.</p>
```
```suggestion
<p>For specific cases where it is seen as beneficial, a separate <strong>Formal syntax</strong> section may be declared using the formal notification.</p>
```

Elchi3 reviewed (changes requested):
Thanks, great to see us moving in this direction 👍 Some changes requested.

Elchi3 commented:
Oh, one more thing: this needs an entry in the MDN contribution changelog.

hamishwillee commented:
Thanks for the review. I will revisit this after people have had a chance to comment on https://github.com/mdn/content/pull/2376#discussion_r578840238

@chrisdavidmills says he will take care of the contribution changelog - presumably after this goes in.

chrisdavidmills commented:
> Thanks for the review. I will revisit this after people have had a chance to comment on [#2376 (comment)](https://github.com/mdn/content/pull/2376#discussion_r578840238)
> 
> @chrisdavidmills says he will take care of the contribution changelog - presumably after this goes in.

I've submitted a draft PR for this at https://github.com/mdn/content/pull/2475

hamishwillee committed:
Fix the single line cases

hamishwillee commented on files/en-us/mdn/structures/syntax_sections/index.html:
```diff
@@ -16,66 +16,93 @@
+<pre class="brush: js">new IntersectionObserver(<var>callback</var>, <var>options</var>)</pre>
 
-<p>or this (see {{DOMxRef("WindowOrWorkerGlobalScope.fetch")}}):</p>
+<p>or this (from {{DOMxRef("Document.hasStorageAccess")}}):</p>
```
FYI this replaces the mixin case. It isn't very interesting though.

hamishwillee commented on files/en-us/mdn/structures/syntax_sections/index.html:
```diff
@@ -87,16 +114,17 @@ <h5 id="Concise_syntax_blocks">Concise syntax blocks</h5>
 <p>But this version is much more concise, and doesn't include the superfluous {{JSxRef("Promise.prototype.then()")}} method call:</p>
 
-<pre class="brush: js"><var>promiseResponse</var> = <var>caches</var>.match(<var>request</var>, <var>options</var>);</pre>
+<pre class="brush: js">match(<var>request</var>, <var>options</var>)</pre>
```
FYI it feels like for the single line case at least we're losing a little information. For example here we now no longer know that a promise is returned.

hamishwillee commented:
@Elchi3 Should be good for a re-review now. The main update being consistent application of the same style to the single-line case.

peterbe commented on files/en-us/mdn/guidelines/writing_style_guide/index.html:
```diff
@@ -236,7 +236,7 @@ <h4 id="Tabs_and_line_breaks">Tabs and line breaks</h4>
 <h4 id="Inline_code_formatting">Inline code formatting</h4>
 
-<p>Use the "Code" button (labeled with two angle brackets "<strong>&lt;&gt;</strong>") to apply inline code-style formatting to function names, variable names, and method names. (This uses the {{HTMLElement("code")}} element). For example: "the <code class="js plain">frenchText()</code> function".</p>
+<p>Use the {{HTMLElement("code")}} tags to mark up function names, variable names, and method names. For example: "the <code>frenchText()</code> function".</p>
```
I don't think that's an appropriate hyperlink. (referring to the use of the `HTMLElement`). You just need to say:
```suggestion
<p>Use the &lt;code&gt; tags to mark up function names, variable names, and method names. For example: <code>the &lt;code&gt;frenchText()&lt;/code&gt; function</code>.</p>
```
(man, this would be easier if it was just Markdown)

hamishwillee replied:
Fair enough. Fixed. Yes, bring on the markdown!

peterbe commented on files/en-us/mdn/structures/syntax_sections/index.html:
```diff
@@ -16,66 +16,93 @@
+<pre class="brush: html">
+  &lt;h2 id="Syntax"&gt;Syntax&lt;/h2&gt;
+
+  &lt;pre class="brush: js"&gt;slice()
```
[All statements must end with semicolons (";").](http://localhost:3000/en-US/docs/MDN/Guidelines/Code_guidelines/JavaScript#general_javascript_guidelines)
So the 3 lines of JS in this example should have a trailing semi-colon.

Ryuno-Ki replied:
Nit: there's Automatic Semicolon Injection (ASI) in JavaScript.
So „should”, not „must” :)

peterbe replied:
I don't know what you're trying to add with that comment @Ryuno-Ki 
This JS code is sample JS code about writing JS in code snippets. There exists a best practice document. So in a sense, the right word is "must" :)

Ryuno-Ki replied:
> All statements must end with semicolons (";"). We require them in all of our code samples even though they're technically optional in JavaScript because we feel that it leads to code that is clearer and more precise about where each statement ends.

I was referring to the „technically optional” bit.

hamishwillee replied:
FWIW I did think about this and had _decided_ to omit them. But worth keeping in line with guidelines. Fixed.

peterbe reviewed (changes requested):
I'm a bit out of my element here. I'm not confident enough on the "policies" for the content and stuff. Sorry. 

However, found some other nits. 

Also, I'm hesitant with this use of `<var>`. It will entirely disappear in the final rendered HTML. The Yari builder will look at the `<pre>` tag much like a user does when using a mouse cursor. It just takes the visible text (none of the markup) and sends it into `PrismJS` for conversion to syntax highlighted HTML. 

And once we move to Markdown, I'm not sure but I suspect we'll have even less power over the syntax-highlighted code blocks. I do hope we can someday have something for code snippets so you can extra highlight specific line numbers.

hamishwillee committed:
No vars, semicolon statements

hamishwillee commented:
@peterbe  ( and @Ryuno-Ki ) Thanks for your reviews

> Also, I'm hesitant with this use of `<var>`. It will entirely disappear in the final rendered HTML. The Yari builder will look at the <pre> tag much like a user does when using a mouse cursor. It just takes the visible text (none of the markup) and sends it into PrismJS for conversion to syntax highlighted HTML.

I have removed all the `var`. While your arguments about yari and markdown are all compelling, the main reason is that they no longer serve a purpose - we've removed all the bits that this was supposed to differentiate. All that is left is the arguments, and they are "by definition" variable/user-defined.

hamishwillee commented:
Self merging following approvals. FYI @chrisdavidmills we can iterate if you have further comments.

I have one; I would love a return type :-)

chrisdavidmills commented:
@hamishwillee nah, this is good; I'm happy with it as it stands. I'm still not convinced about the return type in the syntax...
